reyna/src/com/b2msolutions/reyna/Repository.java

```diff
@@ -12,144 +12,269 @@
+    public static final int DATABASE_VERSION = 1;
 
-	public Repository(Context context) {
-		super(context, DATABASE_NAME, null, DATABASE_VERSION);
+    private static final String TAG = "Repository";
+
+    public Repository(Context context) {
+        super(context, DATABASE_NAME, null, DATABASE_VERSION);
     }
 
-	@Override
-	public void onCreate(SQLiteDatabase db) {
-		Logger.v(TAG, "onCreate");
-		db.execSQL("CREATE TABLE Message (id INTEGER PRIMARY KEY AUTOINCREMENT, url TEXT, body TEXT);");
-		db.execSQL("CREATE TABLE Header (id INTEGER PRIMARY KEY AUTOINCREMENT, messageid INTEGER, key TEXT, value TEXT, FOREIGN KEY(messageid) REFERENCES message(id));");
-	}
+    @Override
+    public void onCreate(SQLiteDatabase db) {
+        Logger.v(TAG, "onCreate");
+        db.execSQL("CREATE TABLE Message (id INTEGER PRIMARY KEY AUTOINCREMENT, url TEXT, body TEXT);");
+        db.execSQL("CREATE TABLE Header (id INTEGER PRIMARY KEY AUTOINCREMENT, messageid INTEGER, key TEXT, value TEXT, FOREIGN KEY(messageid) REFERENCES message(id));");
+    }
 
-	@Override
-	public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+    @Override
+    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
         Logger.v(TAG, "onUpgrade");
-	}
+    }
 
-	public void insert(Message message) {
+    public void insert(Message message) {
         Logger.v(TAG, "insert");
-		if (message == null)
-			return;
-
-		SQLiteDatabase db = null;
-		try {
-			db = this.getWritableDatabase();
-			db.beginTransaction();
-			ContentValues values = new ContentValues();
-			values.put("url", message.getUrl());
-			values.put("body", message.getBody());
-
-			long messageid = db.insert("Message", null, values);			
-			this.addHeaders(db, messageid, message.getHeaders());
-			db.setTransactionSuccessful();
-
-            Logger.i("reyna", "Repository: inserted message " + messageid);
-		} finally {
-			if (db != null) {
-				db.endTransaction();
-			}
-		}
-	}
-
-	public Message getNext() throws URISyntaxException {
+        if (message == null) {
+            return;
+        }
+
+        SQLiteDatabase db = this.getWritableDatabase();
+        this.insertMessage(db, message);
+    }
+
+    public void insert(Message message, long dbSizeLimit) {
+        Logger.v(TAG, "insert with limit");
+        if (message == null) {
+            return;
+        }
+
+        SQLiteDatabase db = this.getWritableDatabase();
+        long dbSize = this.getDbSize(db);
+
+        if (this.dbSizeApproachesLimit(dbSize, dbSizeLimit)) {
+            this.clearOldRecords(db, message);
+        } else if (dbSize > dbSizeLimit) {
+            this.shrinkDb(db, dbSizeLimit, dbSize);
+        }
+
+        this.insertMessage(db, message);
+    }
+
+    private boolean dbSizeApproachesLimit(long dbSize, long limit) {
+        return (limit > dbSize) && (limit - dbSize) < (limit * 0.2); // less than 20% is considered as close to threshold
```

Comment:
do we need to include @Anton suggestion to make it fix the threshold to be 300KB instead of percentage. 
it will be better for storage limit 10MB.
@Youhana-Hana